posts/2020-10-14-lock-poisoning-survey.md

```diff
@@ -0,0 +1,129 @@
+
+The Libs team is looking at how we can improve the `std::sync` module, by potentially splitting it up into new modules and making some changes to APIs along the way.
+One of those API changes we're looking at is non-poisoning implementations of `Mutex` and `RwLock`.
+To find the best path forward we're conducting a survey to get a clearer picture of how the standard locks are used out in the wild.
+
+The survey is a Google Form.
+[You can fill it out here](https://docs.google.com/forms/d/e/1FAIpQLSehk-GkwoCag_w3YfXDfgeANulR0h5m2d3EzUMQaiY1vRfIEw/viewform).
+
+### What is this survey for?
+
+The survey is intended to answer the following questions:
+
+- When is poisoning on `Mutex` and `RwLock` being used deliberately.
+- Whether `Mutex` and `RwLock` (and their guard types) appear in the public API of libraries.
+- How much friction there is switching from the poisoning `Mutex` and `RwLock` locks to non-poisoning ones (such as from `antidote` or `parking_lot`).
+
+This information will then inform an RFC that will set out a path to non-poisoning locks in the standard library.
+It may also give us a starting point for looking at the tangentially related `UnwindSafe` and `RefUnwindSafe` traits for panic safety.
+
+### Who is this survey for?
+
+If you write code that uses locks then this survey is for you.
+That includes the standard library's `Mutex` and `RwLock` as well as locks from `crates.io`, such as `antidote`, `parking_lot`, and `tokio::sync`.
+
+### So what is poisoning anyway?
+
+Let's say you have an `Account` that can update its balance:
+
+```rust
+impl Account {
+    pub fn update_balance(&mut self, change: i32) {
+        self.balance += change;
+        self.changes.push(change);
+    }
+}
+```
+
+Let's also say we have the invariant that `balance == changes.sum()`.
+We'll call this the _balance invariant_.
+So at any point when interacting with an `Account` you can always depend on its `balance` being the sum of its `changes`, thanks to the balance invariant.
+
+There's a point in our `update_balance` method where the balance invariant isn't maintained though:
+
+```rust
+impl Account {
+    pub fn update_balance(&mut self, change: i32) {
+        self.balance += change;
+//      self.balance != self.changes.sum()
+        self.changes.push(change);
+    }
+}
+```
+
+That seems ok, because we're in the middle of a method with exclusive access to our `Account` and everything is back to good when we return.
+There isn't a `Result` or `?` to be seen so we know there's no chance of an early return before the balance invariant is restored. Or so we think.
+
+What if `self.changes.push` didn't return normally?
+What if it panicked instead without actually doing anything?
+Then we'd return from `update_balance` early without restoring the balance invariant.
+That seems ok too, because a panic will start unwinding the thread it was called from, leaving no trace of any data it owned behind. No data means no broken invariants.
```

Comment:
Should this mention that `drop()` might have to deal with broken invariants?

Reply:
It might be a bit in the weeds, but maybe just a tiny note at the end of _No data means no broken invariants_ would be enough?

Reply:
Sure, maybe something like
```suggestion
That seems ok too, because a panic will start unwinding the thread it was called from, leaving no trace of any data it owned behind. No data means no broken invariants (except for `Drop` implementations).
```

Reply:
Maybe we could be even more mysterious and say something like _(the `Drop` trait actually has some interesting implications here)_ to invite the reader to explore what those implications might be without diverting too much?